### ISSUE TITLE
Integrate with WP to allow permissions based dashboards

### ISSUE THREAD

**Issue by key2market** [key2market/dataviz#31]
Could we also integrate this with the WP to allow each user have their own Visualizations and DAshboards and also make sure that other users do not see each other's data and settings?

**Comment by key2market** [comment]
We might want to manage it via MySQL table(s) to alter address things like:
- what is user wants to share dashboard with other users? Perhaps we can do shareable unique link
- what if user creates a dashboard that many other users will want to view.
- Could we add rights management, such as view only permission etc?

**Comment by prahladyeri** [comment]
Yep. I think the easiest way to achieve this is to create a simple WordPress page and call `add_shortcode` to redirect it to the `viz` app with the current user. The app will then do the rendering of dash based on the received querystring. I'll soon send you the updated code.

Reference:
http://wordpress.stackexchange.com/questions/105306/menu-link-redirect-based-on-username-or-user-id

**Comment by prahladyeri** [comment]
Here are the documented steps to achieve this (once I update the dash functionality to make it happen):

- Open the `functions.php` of you main theme file and paste this code anywhere in the beginning:
```

/*Added by Prahlad for vizdata functionality*/
add_action('init', 'add_viz_shortcode');
//wpse105306_add_shortcode();
function add_viz_shortcode()
{
    add_shortcode('viz_user', 'viz_user');
    //error_log('added shortcode.');
}

function viz_user()
{
	//error_log("VIZ_SHORTCODE");
    $user = wp_get_current_user();
    $uid = "uid_" . $user->ID;
    //wp_redirect( 'http://www.key2market.com/dash.html?tag=' . $uid, 301 ); 
    return $uid;
    //exit;
    // do stuff with $projects
}
/*Added by Prahlad for vizdata functionality*/

```
- Once you do that, use the following url link anywhere inside your WP pages/posts to redirect the current user to his/her dash:

`http://www.key2market.com/viz/dash.html?tag=[viz_user]`

**Comment by key2market** [comment]
Ok, Please make sure it is "hack" proof so that user cannot just emulate another user by entering user_id into URL or something like that. This needs to be relatively secure so clients can have confidence that their data is not seen by someone else.

**Comment by key2market** [comment]
?tag=[viz_user] and what if I put random viz user ID here?

**Comment by prahladyeri** [comment]
Then also it will work, all the viz knows is to filter and group charts according to the tag it receives.

**Comment by key2market** [comment]
Can we make it a bit more fool proof?

**Comment by prahladyeri** [comment]
Fool proof means one user cannot access another's dash?  Sure, I'll encrypt the ID while sending from WP and decrypt back inside viz, so one user can't access another's. Is that ok?

**Comment by prahladyeri** [comment]
But this isn't done yet. I'll commit the changes to dash in a while after that it will start to work. You can make these changes to the WP though, until new dash is deployed, it will just redirect to normal dash.

**Comment by key2market** [comment]
Why can't we just use WP is_login(check and then fetch the right dashboards?) Dues to server-side, client-side issues?

**Comment by prahladyeri** [comment]
I just checked that when no user is logged in, the ID is generated as 0, so "uid_0" (encrypted, of course) will be passed to data_viz. You can consider it as a common viz session for guest users.

**Comment by key2market** [comment]
great :)

**Comment by prahladyeri** [comment]
@key2market There is a slight change, from you WP page, add redirect to the main `index.html` viz page, not `dash.html`. I've made the relevant correction on this line:

     'http://www.key2market.com/index.html?tag=' . $uid, 301

Once you go to `index.html`, each user will see their own dashes. I haven't added encryption yet, nor have I made the integration for charts, this is presently only for dashboards. Once you confirm and tell me it functions ok, then I'll add encryption and integration for charts too.

**Comment by prahladyeri** [comment]
I've made the WP integration changes in the master branch. Did you check it?
